stats/statsvariables.cpp

```diff
@@ -0,0 +1,1674 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "statsvariables.h"
+#include "statstranslations.h"
+#include "core/dive.h"
+#include "core/divemode.h"
+#include "core/divesite.h"
+#include "core/gas.h"
+#include "core/pref.h"
+#include "core/qthelper.h" // for get_depth_unit() et al.
+#include "core/subsurface-time.h"
+#include <limits>
+#include <QLocale>
+
+#if QT_VERSION >= QT_VERSION_CHECK(5, 14, 0)
+#define SKIP_EMPTY Qt::SkipEmptyParts
+#else
+#define SKIP_EMPTY QString::SkipEmptyParts
+#endif
+
+static const constexpr double NaN = std::numeric_limits<double>::quiet_NaN();
+
+// Typedefs for year / quarter or month binners
+using year_quarter = std::pair<unsigned short, unsigned short>;
+using year_month = std::pair<unsigned short, unsigned short>;
+
+// Small helper template: add an item to an unsorted vector, if its not already there.
+template<typename T>
+static void add_to_vector_unique(std::vector<T> &v, const T &item)
+{
+	if (std::find(v.begin(), v.end(), item) == v.end())
+		v.push_back(item);
+}
+
+// Small helper: make a comma separeted list of a vector of QStrings
+static QString join_strings(const std::vector<QString> &v)
+{
+	QString res;
+	for (const QString &s: v) {
+		if (!res.isEmpty())
+			res += ", ";
+		res += s;
+	}
+	return res;
+}
+
+// Note: usually I dislike functions defined inside class/struct
+// declarations ("Java style"). However, for brevity this is done
+// in this rather template-heavy source file more or less consistently.
+
+// Templates to define invalid values for variables and test for said values.
+// This is used by the binners: returning such a value means "ignore this dive".
+template<typename T> T invalid_value();
+template<> int invalid_value<int>()
+{
+	return std::numeric_limits<int>::max();
+}
+template<> double invalid_value<double>()
+{
+	return std::numeric_limits<double>::quiet_NaN();
+}
+template<> QString invalid_value<QString>()
+{
+	return QString();
+}
+template<> StatsQuartiles invalid_value<StatsQuartiles>()
+{
+	double NaN = std::numeric_limits<double>::quiet_NaN();
+	return { NaN, NaN, NaN, NaN, NaN };
+}
+
+static bool is_invalid_value(int i)
+{
+	return i == std::numeric_limits<int>::max();
+}
+
+static bool is_invalid_value(double d)
+{
+	return std::isnan(d);
+}
+
+static bool is_invalid_value(const QString &s)
+{
+	return s.isEmpty();
+}
+
+// Currently, we don't support invalid dates - should we?
+static bool is_invalid_value(const year_quarter &)
+{
+	return false;
+}
+
+static bool is_invalid_value(const dive_site *d)
+{
+	return !d;
+}
+
+static bool is_invalid_value(const StatsOperationResults &res)
+{
+	return !res.isValid();
+}
+
+struct gas_bin_t {
+	enum class Type {
+		Air,
+		Oxygen,
+		EAN,
+		Trimix
+	} type;
```

Comment:
so three commits ago you added the gas types. There you didn't have **oxygen**, but you had **normoxic** (and were IMHO also missing **hypoxic**)
Maybe that's a different use case, but that would be at least worth a comment, right?

Reply:
Yes, the former was for binning with a "general type". This is for formatting bins that use the actual composition of the gases. Which is fun, because it means that we have two-dimensional bins. I shortly entertained the idea, but then refrained from doing 3D histograms (O2 vs H2 vs Count). 😀